### PR TITLE
Run and show test results in the editor

### PR DESCRIPTION
**Pull request by jacobmendoza**
Overview: Prototype for integrating a runner of AVA tests in Atom. Not ready to merge, just requesting feedback (https://github.com/sindresorhus/atom-ava/issues/3).

Steps: Link the package. At this stage, there is no menu integration (yet), so pressing ctrl-alt-a is needed to activate the package (right pane will appear). If a test file is active, pressing ctrl-alt-x will trigger the execution of the tests. Every time that one assert is processed, a new line will appear in the UI (the feedback during execution needs to be improved).

About the code: It's just a prototype, so probably there are a lot of things to discuss. Intentionally, there is no effort in this PR for formatted and nice UI. I'm using this [parser](https://github.com/tapjs/tap-parser) for the TAP output, and seems to work very well. Anyway, the idea is to abstract the interaction with it, just in case we could need to replace it (with this code, this is only partially achieved).

Part of the pageres suite

### CODE REVIEW

**Comment by sindresorhus** [comment]
Good start! I know we didn't discuss this upfront, but the plugin should use ES2015 (ES6), not CoffeeScript. Nothing against CoffeeScript, but ES2015 is more [future friendly](https://twitter.com/sindresorhus/status/669958889941639168) and better for attracting potential contributors.

> At this stage, there is no menu integration (yet)

What kind of menus were you thinking? I think it's more important to implement command palette commands than menus.

File: keymaps/ava.cson [hunk 1]
```diff
@@ -0,0 +1,3 @@
+'atom-workspace':
+  'ctrl-alt-a': 'ava:toggle'
+  'ctrl-alt-x': 'ava:run'
```
Comment (sindresorhus) [inline keymaps/ava.cson]: Is there really any point in having two separate shortcuts? I would think it's enough with just Ctrl+Alt+A which toggles and runs tests.

File: lib/test-runner-process.coffee [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    run: (filePath, assertCallback, completeCallback) ->
+      @parser = @getParser()
+      @parser.on('assert', assertCallback)
+      @parser.on('complete', completeCallback)
+
+      #TODO: Fix parsing of folders and files
+      folder = filePath.substring(0, filePath.lastIndexOf("/") + 1);
+      file = filePath.split("/").pop()
+
+      command = "ava #{file} --tap"
```
Comment (sindresorhus) [inline lib/test-runner-process.coffee]: Would be better to use the AVA programmatic API. Example: https://github.com/pine613/fly-ava/blob/master/index.js

File: package.json [hunk 3]
```diff
@@ -35,6 +39,10 @@
     "ava",
     "mocha"
   ],
+  "dependencies": {
+    "event-kit": "^1.1.0",
+    "tap-parser": ">=v1.2.2"
```
Comment (sindresorhus) [inline package.json]: `>=v1.2.2` => `^1.2.2`

**Comment by jacobmendoza** [comment]
I wasn't aware of [this](http://blog.atom.io/2015/02/04/built-in-6to5.html) but these are actually good news. No doubt this is more future friendly and has more potential. I'll start doing some testing.

The comment about the menu did not have a big foundation, I was just trying to say that there is no visual entry point in the UI. Probably we don't need it.

File: keymaps/ava.cson (hunk 1 above)
Comment (jacobmendoza) [inline keymaps/ava.cson]: There is no big reason for having two today. I used two because they allowed me to easily test in isolation the two steps. I'll compact them in one. 

If in the future we have a parsing of the test file (something that I'd like to discuss after having a nice stable first version) phase we'll need two. No need to worry about this now.

File: lib/test-runner-process.coffee (hunk 2 above)
Comment (jacobmendoza) [inline lib/test-runner-process.coffee]: Of course! Will do!.

**Comment by jacobmendoza** [comment]
Thanks for all the feedback :). I'll try to push a new version with these issues addressed asap.

File: lib/test-runner-process.coffee (hunk 2 above)
Comment (jacobmendoza) [inline lib/test-runner-process.coffee]: @sindresorhus, sorry if I'm missing something here.

When trying to use the programmatic API providing an absolute path for running a test I'm getting 'No tests found in ../x/y.js, make sure to import "ava" at the top of your test file'. Same result if I use the CLI in a folder different from the one that hosts the file, as: 
- ava /Users/Me/Folder/test.js --verbose

I have been able to trace the code till test-worker.js but being honest, I don't know why it's failing. Any idea?. Maybe I'm doing something wrong...

Thank you! :)

**Comment by jacobmendoza** [comment]
I have addressed some of the issues previously marked:
- The files are now ES6.
- Dependencies fixed.
- I'm still not using the programmatic API from AVA. **Pending**.
- Ctrl+A toggles the package and runs the tests.
- I haven't done anything with the UI. At this stage is only showing the tests. There is no method to show errors.

I have tried to implement a non very successful cancellation mechanism. I'm not sure if it's the best approach, but the idea is stop the process if the user closes the package or if the test execution process is launched several times. It'd great have some feedback about the appropriate way to do it with the programmatic API.

Is it looking any better?. Thank you!.

File: lib/main.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+'use babel';
```
Comment (sindresorhus) [inline lib/main.js]: Use `/** @babel */` instead.

File: package.json [hunk 5]
```diff
@@ -35,10 +39,15 @@
...
   ],
+  "dependencies": {
+    "tap-parser": "^1.2.2",
+    "ava": "^0.13.0"
+  },
   "devDependencies": {
     "xo": "*"
   },
   "xo": {
+    "envs": ["browser", "node", "jasmine", "atomtest"],
```
Comment (sindresorhus) [inline package.json]: One item on a separate line

File: lib/test-runner-process.coffee (hunk 2 above)
Comment (sindresorhus) [inline lib/test-runner-process.coffee]: AVA 0.13.0 changed the API regarding how files are added. Ensure you pass them to the `.run()` method: https://github.com/sindresorhus/ava/blob/6d5f3225cc712f10e7a0a8a9c6e40d574fd4b251/test/api.js#L23-L27

> Same result if I use the CLI in a folder different from the one that hosts the file, as:

I can reproduce too. Would you mind opening an issue on AVA?

```
~/dev/chalk master
❯ ava /Users/sindresorhus/dev/acosh/test.js

  ✖ No tests found in ../acosh/test.js, make sure to import "ava" at the top of your test file
```

## 

As a dirty workaround, for now you could probably do this https://github.com/sindresorhus/atom-linter-xo/commit/5d85618ad695b9b7142ac4015a6dde6d8709c6f9.

File: keymaps/ava.cson [hunk 6]
```diff
@@ -0,0 +1,2 @@
+'atom-workspace':
+  'ctrl-alt-a': 'ava:toggle'
```
Comment (sindresorhus) [inline keymaps/ava.cson]: Can you use a JSON file here instead?

File: lib/main.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+const Panel = require('./panel');
+const {CompositeDisposable} = require('atom');
```
Comment (sindresorhus) [inline lib/main.js]: Since it goes through Babel, you can use ES2015 `import` syntax.

File: package.json [hunk 8]
```diff
@@ -2,6 +2,7 @@
   "name": "ava",
   "version": "0.2.0",
   "description": "Snippets for AVA",
+  "main": "./lib/main",
```
Comment (sindresorhus) [inline package.json:5]: Just rename `main.js` to `index.js` and use `"main": "lib",`

File: spec/main-spec.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
+'use babel';
+
+describe('TestingForAva', () => {
```
Comment (sindresorhus) [inline spec/main-spec.js:3]: Do we have to use Jasmine for testing Atom packages or could we use AVA?

File: styles/ava.less [hunk 10]
```diff
@@ -0,0 +1,11 @@
+@import "ui-variables";
```
Comment (sindresorhus) [inline styles/ava.less]: single-quotes

File: styles/ava.less [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+.ava {
+  padding: 30px;
+  font-size: 15px;
+
+  .message { padding-bottom: 10px; }
+  .summary { font-size: 12px; padding-top: 10px; }
+  .test { font-size: 11px; font-weight: bold; }
+  .executing { font-size: 12px; font-weight: bold; padding-bottom: 10px;}
```
Comment (sindresorhus) [inline styles/ava.less]: use tab indentation

**Comment by sindresorhus** [comment]
Nice progress :) Added some more feedback.

File: lib/test-runner-process.coffee (hunk 2 above)
Comment (jacobmendoza) [inline lib/test-runner-process.coffee]: No worries, I will open the issue!. I thought that it was more likely to be something wrong with my environment. I saw the change in the API and was providing the files to the run() method. 

I'll take a look at the workaround. Thanks :).

File: package.json (hunk 8 above)
Comment (jacobmendoza) [inline package.json:5]: Not working for me... not sure if I'm missing something.

File: spec/main-spec.js (hunk 9 above)
Comment (jacobmendoza) [inline spec/main-spec.js:3]: Yeah, is more consistent and makes a lot of sense use AVA. I have been thinking about that since the beginning. My only doubt is if it can lead to any problems, just because Jasmine in the de facto option in Atom. 

Probably is time to make some tests.

**Comment by jacobmendoza** [comment]
@sindresorhus, just wanted to check if you have any ideas about something that came up while thinking about the plugin and the programmatic api. 

At least speaking about the plugin I think makes sense have a mechanism that allows users to cancel the current execution of the suite. Moreover, in the plugin context, that could control some edge cases, as the user pressing several times the key that runs the tests and things like that.

Any thoughts about that?. Is there any way to get that with the programmatic api?. I understand that it must be difficult due to the async nature of ava, but just in case.

If this makes more sense in gitter or as an issue let me know!.

Thanks,

**Comment by sindresorhus** [comment]
> At least speaking about the plugin I think makes sense have a mechanism that allows users to cancel the current execution of the suite.

That makes sense. Not something we currently support though. https://github.com/sindresorhus/ava/issues/710

> Moreover, in the plugin context, that could control some edge cases, as the user pressing several times the key that runs the tests and things like that.

That can easily be worked around for now by not allowing a new run when one is in progress.

## 

Feel free to continue using the CLI if you think it would be better for now.

**Comment by jacobmendoza** [comment]
Yeah, after testing a bit I think I'll continue with the CLI (for now) and will apply the workaround as suggested. I think that makes sense.

Progress soon. 

Thanks @sindresorhus!,

**Comment by jacobmendoza** [comment]
1. Initial version of the UI. Deliberately, just for now, I have used DOM manipulation. I know that it does not look good. There is an ongoing discussion in GitHub about a view framework for the packages (https://github.com/atom/atom/issues/5756) and I need to take a look.
2. For now we only have the UI being rendered in a panel. When more options appear (block decorations or mini-version in the toolbar, etc.) I think it should be easy to extend after some refactor. For getting something done about that, I will need also a correlation between the result that I get and the line that the test occupies in the file. Will try to get something done about that.
3. Also, for now, it is only executing the current file, not the entire project. I think a decision about a view framework should be made before implementing this. There are two commands, run (ctrl-alt-r) and toggle (ctrl-alt-a). The toggle is there because is handy hide the panel with the keystroke.

screen shot 2016-04-30 at 02 20 17

**Comment by jamestalmage** [comment]
Oooh! You are tempting me to switch from WebStorm with this!

**Comment by jacobmendoza** [comment]
@jamestalmage, lots of things to do yet, some embarrassing enough parts in the code, but hopefully it will get better. I appreciate a lot your comment :).

**Comment by jacobmendoza** [comment]
@sindresorhus, @jamestalmage, if you have time it would be great if you can give it a go and let me know what you think:

[Video](https://www.youtube.com/watch?v=ARfSwc7V61Q&feature=youtu.be)
(When the video was recorded both columns had 'passed' as name. It has been fixed)
screen shot 2016-05-02 at 17 39 33

Some minor changes in the UI to prepare the plugin for execute several files. The idea is that we will have several blocks as the one featured in the video (one per file).

My initial understanding is that some of the features that I feel like missing (block decoration and multiple files) will require some minor modifications in the TAP reporter of AVA, that is why I would like to know if the features that we have make sense as an MVP before going for that second phase.

Thanks!

**Comment by sindresorhus** [comment]
@jacobmendoza This looks really great! 🙌 More than good enough for a MVP.

## 

Some minor nitpick. When using the Atom Dark theme some things could be tweaked:

screen shot 2016-05-04 at 22 03 11

Maybe use `rgba()` with lowered opacity on the horizontal line and the `test.js` background. Or maybe there's a built-in theme background variable we could use.

And `✖` instead of the current `x` for closing the sidebar.

## 

And I get this when trying to run a test file from the [`got`](https://github.com/sindresorhus/got) project. Maybe because it's in a nested folder? I'm running the command directly from the file.

screen shot 2016-05-04 at 22 10 59

**Comment by jacobmendoza** [comment]
Hey @sindresorhus, thanks for spending some time with this!.

1) I had in mind the style for the dark theme and I forgot in the last minute. Yeah, we should have some variables in place that should help us. I'll take a look.

2) I've changed the close icon already. Will push the commit with all the other changes.

3) That last error that you have there is strange!. I have cloned the project and tried to reproduce. I have opened atom in the project folder, and run the file (gzip.js) with the command and with the keyboard shortcut. Worked for me with both of them, also for the other files of the project. Is it happening just for that one?.

**Comment by jamestalmage** [comment]
Very cool.

Only thing I've noticed so far is that skipped tests show as passed.

**Comment by jfmengels** [comment]
> Only thing I've noticed so far is that skipped tests show as passed.

And `todo` tests are shown as failed.

Looks like a great start! Great job @jacobmendoza!

**Comment by jacobmendoza** [comment]
Hey @jamestalmage, @jfmengels!.

Thanks for trying out the package! :). Yeah, that's important, something funny with the parsing process I'd say. Hopefully it'll be fixed very soon.

**Comment by jamestalmage** [comment]
@jacobmendoza 
If it's a parsing problem, that could very well be AVA's fault. Our TAP output isn't super sophisticated. 

https://github.com/jamestalmage/tap-emitter aims to fix that, but I've kind of let that languish. I would love extra help on that. It would require fixing https://github.com/jamestalmage/tap-emitter/issues/2 and integrating into AVA. No pressure, but it might make your life easier to improve the TAP output.

**Comment by jacobmendoza** [comment]
@jamestalmage, not sure yet if it's the output that AVA generates or actually the parsing process. I think I'll be able to confirm today.

Anyway, yeah, I would be more than happy to help. This is pretty far from my usual technology stack, so I may need some help for start. Actually, regarding the TAP output, there are a few ideas that I wanted to check with you, but I was trying to get something basic working here before.

I'll keep you posted!

**Comment by jacobmendoza** [comment]
@jamestalmage, @jfmengels, thanks for your comments yesterday. They were really helpful.

1) About running tests that were marked as skip/todo, should be working now. AVA is doing just fine, and literally, it wasn't implemented (6d6449a - edit: just noticed a style defect there). For now, they just won't appear in the summary and statistics and they will appear in the list in a different colour. If you have an opinion about the presence/shape that they should have in the UI let me know!. Used the most basic approach for now. 

2) @jamestalmage, again, there are more things that I'd like to get from the TAP reporter, but my lack of knowledge is so big that I don't even know if my ideas make sense or what would be the best approach to implement them. I'll contact you after putting this in a reasonable state for a first phase.

3) The plugin should be looking significantly better in different themes. I've tried with 6 and at least the result is reasonable. I'm using some pre-defined variables of the themes (7d3aa5e).

4) @sindresorhus, I've just realised. In the package [I'm making reference to the AVA logo directly in the AVA repo](https://github.com/sindresorhus/atom-ava/blob/runner-start/views/panel.html#L3). If the network fails or is slow, it will look pretty funny. Can I copy it to this package?. The only problem that I see is that we will have two places to maintain the logo if it changes, but even with that, I guess that it should be a local asset of the plugin.

Also, fixed a crash if the user tries to run tests in a window with no active editor.

**Comment by sindresorhus** [comment]
> Can I copy it to this package?

Yes, should be in the package.

**Comment by jacobmendoza** [comment]
@sindresorhus, did you take a look at the error that you were having while trying to test [got](https://github.com/sindresorhus/got)?. Can you still reproduce it?.

I was wondering if, assuming that the defect is no longer present, we could start thinking about merging with the main branch. This PR is starting to get big, and I would like to make a difference between the first basic features and the followings to come.

Thank you!

**Comment by sindresorhus** [comment]
@jacobmendoza Yes, works great now! This is really awesome. We really appreciate all the hard work you put into this.

superhighfive

**Comment by sindresorhus** [comment]
I'll do a new release and tweet about it next week ;)

**Comment by jfmengels** [comment]
fmkp70c7IIb9C

**Comment by jacobmendoza** [comment]
@sindresorhus, @jfmengels,

Thanks a million for your kind words, really ✋✋✋ . I have enjoyed it a lot and it has been great to have your feedback during the process!.

Planning to comment soon https://github.com/sindresorhus/atom-ava/issues/13 to see what you think.

Cheers guys!
